### DISCUSSION TITLE
Discussion on difference between zkSync and Arbitrum

### DISCUSSION THREAD

**Discussion by Eshan-Sharma** [Cyfrin/foundry-full-course-cu#1920] [General]
https://updraft.cyfrin.io/courses/blockchain-basics/basics/why-zksync-is-our-recommended-chain?lesson_format=video

In the last part of the video, I understand that zkSync is proving the rollups mathematically and is using solidity with a couple of nuances, but did not understand the fundamental difference between Arbitrium's Optimistic rollups and zkRollups. 

I am confused can someone help me understand the difference between the two and why is Arbitrum having more locked value.
Is there anything like foundry -zkSync for Arbitrum as well?

**Comment by EngrPips** [comment]
Hey @Eshan-Sharma, This is a phenomenal question, and I am convinced I won't do justice to the answer if I attempt to answer, so I will say let's humbly wait for @PatrickAlphaC  answer whenever they are available.

**Reply by PatrickAlphaC** [reply]
Great question! I've got Ciara herself who is going to answer!

**Reply by EngrPips** [reply]
Looking forward to it. My note and pen are ready, hahaha.

**Reply by jeka0203** [reply]
Wow, cool

**Reply by EngrPips** [reply]
Indeed

**Comment by ciaranightingale** [comment]
Hey @Eshan-Sharma (and @EngrPips!), fantastic question! 

As you correctly identified:
- zkSync is a zero-knowledge (ZK) rollup 
- Arbitrum is an Optimistic rollup 

The fundamental difference between ZK rollups and optimistic rollups lies in how the transactions are validated (checked to ensure that nothing malicious or fraudulent is happening).
- **zk rollups use mathematical zk proofs** to check the transactions
- **Optimistic rollups assume the transactions are valid by default** but have a period of time in which operators can challenge the transactions (challenge the blockchain state) by sending **fraud-proof.** 

In terms of locked value, Arb has more locked value than zkSync purely because it has been around for longer; therefore, more users have deposited $$ into the system.

This, by the way, would also be a fantastic question to chuck into an AI tool like Chat GPT!:) e.g. "Why does Arbitrum have a higher total value locked (TVL) than zkSync?" 

Additionally,[this article](https://www.cyfrin.io/blog/what-are-blockchain-rollups-a-full-guide-to-zk-and-optimistic-rollups) has a more in-depth breakdown of zk vs optimistic rollups.

Let me know if this helps or if you have any other questions!

**Reply by EngrPips** [reply]
Awesome, Good explanation, @ciaranightingale. You have done justice to this.

**Reply by issa99-lab** [reply]
@EngrPips @Eshan-Sharma Wrote something on the key differences between Optimistic and Zk-rollups. Give it a read [here](https://medium.com/coinmonks/ethereum-layer-2-solutions-optimistic-zk-rollups-1c7c38e6370b) :)

**Reply by EngrPips** [reply]
Thanks for sharing this with us, @issa99-lab. I truly appreciate it.

**Comment by Eshan-Sharma** [comment]
Hey @ciaranightingale ,

Thank you for the explanation and the insightful blog post by @issa99-lab  and yourself—it really helped clarify the underlying mechanisms of zkSync and Arbitrum.

I have a few more questions from a developer's perspective. When developing on the Ethereum network, I typically write my smart contracts in Solidity, validate them with tests, and deploy them using scripts without concerning myself too much with the node operations behind the scenes.

I'm trying to understand if the development process is similar when working with zkSync or Arbitrum. Specifically:

1. Development Process: Is it correct to assume that I can write my contracts, test them, and deploy them on the zkSync testnet or mainnet using Solidity and tools like foundry-zksync, just as I would on Ethereum? Is the process the same for Arbitrum?

2. Triggering Bundling and Validation: For zkSync and Arbitrum, is the bundling and validation of transactions (whether through zk proofs or fraud proofs) something that is handled automatically by the respective rollup chains, or do developers need to trigger these processes manually?

3. EVM Compatibility vs. EVM Equivalence: I came across terms like 'EVM compatible' (zkSync) and 'EVM equivalent' (Arbitrum) in the documentation. While I believe this might not impact the initial contract development phase, could you explain the differences between these concepts? How might these differences affect the later stages of development, particularly in terms of security and auditing?

I want to ensure that I am fully aware of any nuances or differences that could impact my development approach or best practices when working on these platforms.

Thanks again for your guidance and assistance!

**Reply by ciaranightingale** [reply]
Thanks for breaking down the question into steps - super helpful!:)

1. For Arbitrum, the development process is the same as it would be on Ethereum (or Sepolia). The regular (or vanilla as Pat likes to say) version of Foundry can be used. Since zkSync has it's own virtual machine known as the zkSync VM, and functions a little differently, a special fork of Foundry is used (foundry-zksync), this is not the case on Arbitrum and the regular version of Foundry can be used. This is because Arbitrum [is almost EVM equivalent](https://www.evmdiff.com/diff?base=1&target=42161) whereas [zkSync is EVM compatible](https://docs.zksync.io/build/resources/faq#evm-compatibility). 

2. The bundling and validation processes on zkSync occur automatically via the operators. In general, all the developers (or users) need to concern themselves with is sending the transaction (via an RPC) to the network. 

3. Great question! Another question that would be fantastic for an AI assistance as well, if you wanted more details. 

**EVM Equivalence**: These chains offer 100% compliance with the EVM specification. The bytecode and opcodes are identical to those on Ethereum, allowing smart contracts to be deployed without any modifications. 

**EVM Compatibility (e.g., zkSync Era)**: These chains have different bytecode and some opcodes are not supported. For zkSync Era, this means converting Solidity code -> Yul (an intermediate language)->  zkSync VM bytecode using the zksolc compiler (as opposed to the solc compiler for EVM equivalent chains). This is handled by `foundry-zksync`, e.g. we add the `--zksync` flag when building and deploying. So Solidity (and Vyper) both work on zkSync, but the bytecode is completely different. 

So essentially:
- **EVM Equivalence**: the bytecode is the same and all opcodes are supported.
- **EVM Compatibility**: Some of the EVM opcodes are not supported and the bytecode is different (so need a different compiler and a special fork of foundry). For the zkSync differences from Ethereum, [the zkSync docs is a fab resource](https://docs.zksync.io/build/developer-reference/ethereum-differences). 

[Vitalik's blog post](https://vitalik.eth.limo/general/2022/08/04/zkevm.html) is a great resource for this!

In terms of auditing and security, Pat made [this document](https://github.com/Cyfrin/foundry-full-course-cu/blob/main/read-before-deploying-serious-projects-to-zksync.md) on Ts&Cs before deploying a production project to zkSync - @PatrickAlphaC did you have any other thoughts here?

**Reply by EngrPips** [reply]
I am convinced your thoughts are sufficient. Lol. Great breakdown!

**Reply by PatrickAlphaC** [reply]
@ciaranightingale nope, this is great!

**Reply by Eshan-Sharma** [reply]
Thank you @ciaranightingale for the amazing explanation, now my doubts are crystal clear!
